packages/fmg-payments/src/payment-app.ts

```diff
@@ -0,0 +1,142 @@
+import { Commitment, Channel, CommitmentType } from 'fmg-core';
+import { bigNumberify } from 'ethers/utils';
+
+export interface PaymentCommitmentArgs {
+  libraryAddress: string;
+  channelNonce: number;
+  asAddress: string;
+  bsAddress: string;
+  asBalance: string;
+  bsBalance: string;
+  turnNum: number;
+  commitmentCount: number;
+}
+function fromPaymentCommitmentArgs(paymentArgs: PaymentCommitmentArgs) {
+  const {
+    turnNum,
+    asAddress,
+    bsAddress,
+    channelNonce,
+    libraryAddress,
+    asBalance,
+    bsBalance,
+    commitmentCount,
+  } = paymentArgs;
+
+  const channel: Channel = {
+    channelType: libraryAddress,
+    nonce: channelNonce,
+    participants: [asAddress, bsAddress],
+  };
+  return {
+    channel,
+    allocation: [asBalance, bsBalance],
+    destination: [asAddress, bsAddress],
+    commitmentCount,
+    commitmentType: CommitmentType.App,
+    appAttributes: '0x0',
+    turnNum,
+  };
+}
+
+export function preFundSetupCommitment(paymentArgs: PaymentCommitmentArgs): Commitment {
+  return { ...fromPaymentCommitmentArgs(paymentArgs), commitmentType: CommitmentType.PreFundSetup };
+}
+export function postFundSetupCommitment(paymentArgs: PaymentCommitmentArgs): Commitment {
+  return {
+    ...fromPaymentCommitmentArgs(paymentArgs),
+    commitmentType: CommitmentType.PostFundSetup,
+  };
+}
+export function appCommitment(paymentArgs: PaymentCommitmentArgs): Commitment {
+  return { ...fromPaymentCommitmentArgs(paymentArgs), commitmentType: CommitmentType.App };
+}
+export function concludeCommitment(paymentArgs: PaymentCommitmentArgs): Commitment {
+  return { ...fromPaymentCommitmentArgs(paymentArgs), commitmentType: CommitmentType.Conclude };
+}
+
+// Transition helpers
+export function initialPreFundSetup(
+  libraryAddress: string,
+  channelNonce: number,
+  asAddress: string,
+  bsAddress: string,
+  asBalance: string,
+  bsBalance: string,
+): Commitment {
+  return preFundSetupCommitment({
+    libraryAddress,
+    channelNonce,
+    asAddress,
+    bsAddress,
+    asBalance,
+    bsBalance,
+    turnNum: 0,
+    commitmentCount: 0,
+  });
+}
+
+export function nextPreFundSetup(commitment: Commitment): Commitment {
+  return {
+    ...commitment,
+    commitmentCount: commitment.commitmentCount + 1,
+    turnNum: commitment.turnNum + 1,
+  };
+}
+export function postFundSetup(commitment: Commitment): Commitment {
+  // If we are switching from a preFundSetup to a postFundSetup we reset the commitmentCount
+  const commitmentCount =
+    commitment.commitmentType === CommitmentType.PreFundSetup ? 0 : commitment.commitmentCount + 1;
+  return {
+    ...commitment,
+    commitmentType: CommitmentType.PostFundSetup,
+    turnNum: commitment.turnNum + 1,
+    commitmentCount,
+  };
+}
+export function pay(commitment: Commitment, amount: string): Commitment {
```

Comment:
We probably want this to handle the case where it's not our turn and we pay via replacement e.g.

1. playerA sends `{turnNum: 4, aBal: 5, bBal: 5}`
2. playerA sends `{turnNum: 4, aBal: 4, bBal: 6}`

PlayerA can reason that PlayerB will always accept the later state, as the later state is more favourable to PlayerB. Using this 'replace by incentive' strategy, the players can send a stream of payments without a response.

In order to do this, this function probably needs to accept something like `theirAddress` or `ourIndex` or something equivalent as an argument and do the appropriate logic to figure out whether to increase the `turnNum` or not.

Reply:
I've added an `ourIndex` argument to `pay` and it now generates a replacement commitment.